src/Plugin/formatter/Formatter.php

```diff
@@ -81,4 +84,104 @@ protected static function isIterable($input) {
     return is_array($input) || $input instanceof \Traversable || $input instanceof \stdClass;
   }
 
+  /**
+   * Checks if the passed in data to be rendered can be cached.
+   *
+   * @param mixed $data
+   *   The data to be prepared and rendered.
+   *
+   * @return bool
+   *   TRUE if the data can be cached.
+   */
+  protected function isCacheEnabled($data) {
+    // We are only caching field collections, but you could cache at different
+    // layers too.
+    if (!$data instanceof ResourceFieldCollectionInterface) {
+      return FALSE;
+    }
+    if (!$context = $data->getContext()) {
+      return FALSE;
+    }
+    return !empty($context['cache_fragments']);
+  }
+
+  /**
+   * Gets the cached computed value for the fields to be rendered.
+   *
+   * @param mixed $data
+   *   The data to be rendered.
+   *
+   * @return mixed
+   *   The cached data.
+   */
+  protected function getCachedData($data) {
+    if (!$render_cache = $this->createCacheController($data)) {
+      return NULL;
+    }
+    return $render_cache->get();
+  }
+
+  /**
+   * Gets the cached computed value for the fields to be rendered.
+   *
+   * @param mixed $data
+   *   The data to be rendered.
+   * @param mixed $output
+   *   The rendered data to output.
+   */
+  protected function setCachedData($data, $output) {
+    if (!$render_cache = $this->createCacheController($data)) {
+      return;
```

Comment:
In general if you don't have a `@return` tag in your docblock, the recommendation is to return _void_ instead of _NULL_. Even if those are the same in PHP.

Returning different from `return;` in a function that does not return a value will make PhpStorm unhappy.